// routes/adminapi.js
var express = require('express');
var router = express.Router();
var mongoose = require( 'mongoose' );
var multer = require('multer');
var fs = require('fs');
var im = require('imagemagick');
var Post = mongoose.model('Post1212');
var Instagram = mongoose.model('Instagram1212');
var Timetable = mongoose.model('Timetable');
var Video = mongoose.model('Video1');
var Tag = mongoose.model('Tag');
var Ester = mongoose.model('Ester21212');
var SoundCloud = mongoose.model('SoundCloud');
var People = mongoose.model('People');
var Info = mongoose.model('Info');

var storage = multer.diskStorage({
  destination: function (req, file, cb) {
  		console.log("yes1");
  	if(file.fieldname == 'photo_file'){

  		cb(null, './public/img/')
  	}
  	if(file.fieldname == 'songs_files'){
  		cb(null, './public/sound/')
  	}    
  },
  filename: function (req, file, cb) { 
  		console.log("yes2");
  	if(file.fieldname == 'photo_file'){  		
  		if (file.mimetype == 'image/gif'){
  			cb(null, Date.now()+ '.gif')
  		}
  		else {
  			cb(null, Date.now()+ '.png')
  		} 		
  	}
  	if(file.fieldname == 'songs_files'){
  		cb(null, file.originalname + Date.now()+ '.mp3')
  	}    
  }
})


var upload = multer({ storage: storage });
var cpUpload = upload.fields([{ name: 'photo_file', maxCount: 1 }, { name: 'songs_files', maxCount: 10 }]);
var peopleUpload = upload.single('photo_file');

function isAuthenticated (req, res, next) {	
	if(req.method === "GET"){
		return next();
	}
	if (req.isAuthenticated()){
		return next();
	}
	return res.redirect('/login');
};

router.use('/post', isAuthenticated);
router.use('/people', isAuthenticated);
router.use('/timetable', isAuthenticated);
router.use('/instagram', isAuthenticated);
router.use('/soundcloud', isAuthenticated);
router.use('/video', isAuthenticated);
router.use('/tag', isAuthenticated);


router.route('/posts')	
		
	.post(cpUpload, function(req, res) {
		


		var post = new Post();	
		
		post.title = req.body.title;
		post.author = req.body.author;
		post.annotation = req.body.annotation;
		post.content = req.body.content;
		if (req.body.tags != "") {
			post.tags = JSON.parse(req.body.tags);	
		}		
		post.urlname = req.body.urlname;	
		post.type = req.body.type;		
		post.photo = null;
		post.songs = null;
		post.status = req.body.status;
		post.last_mod = new Date();

		if (post.status == "publication") {			
			var ester= new Ester();
			ester.post = post._id;
			ester.type = post.type;
			if (req.body.tags != "") {
				ester.tags = post.tags;	
			}			
			post.ester = ester._id;
		};		

		console.log(post);
		console.log(ester);
		if(req.files) {
			if (req.files['photo_file']) {
				var photo = {};
				photo['url'] = '/img/'+ req.files['photo_file'][0]['filename'];
				photo['path'] = req.files['photo_file'][0]['path'];
				post.photo = photo;
				im.resize({
				  srcPath: photo['path'],
				  dstPath: photo['path'],
				  height:   400,
				  filter: 'Lagrange',
				}, function(err, stdout, stderr){
				  if (err)  console.dir(arguments);
				  im.crop({
					  srcPath: photo['path'],
					  dstPath: photo['path'],
					  width: 400,
					  height: 400,					 
					  
					});
				});
							
			}			
			if (req.files['songs_files']){
				console.log("da");
				console.log(req.files['songs_files']);
				post.songs = [];
				for (var i = 0; i < req.files['songs_files'].length; i++) {
					var song = {
						url : '/sound/'+ req.files['songs_files'][i]['filename'],
						path : req.files['songs_files'][i]['path'],
						name: req.files['songs_files'][i]['originalname']
					};
					post.songs.push(song);					
				};				
			};			
		};
		

		post.save(function(err, post) {
			if (err)
				return res.send(500, err);
			if (post.status == "publication") {				
				ester.save(function(err, ester) {
					if (err)
						return res.send(500, err);					
					return res.send(200, null);
				});
			}
			else 				
				return res.send(200, null);		
		});
	}) 	
	
	


router.route('/posts:title?:offset?')
	

	.get(function(req, res){

		if (req.query.title == "") {
			Post
			.find()
			.sort({created_at: 'desc'})
			.limit(20)
			.skip( 20 * req.query.offset)
			.select('title status created_at last_mod author')			
			.exec(function (err, post) {
				if(err){
					return res.send(500, err);
				}			
				return res.send(200, post);
			});

		}
		else {
			var a = new RegExp(req.query.title,'gi');

			Post
				.find({title: a})
				.sort({created_at: 'desc'})
				.limit(20)
				.skip( 20 * req.query.offset)
				.select('title status created_at last_mod')
				
				.exec(function (err, post) {
					if(err){
						return res.send(500, err);
					}			
					return res.send(200, post);
				});
		}
			
		

	});


router.route('/posts/:id')

	.get(function(req, res){
		if (req.params.id != 'undefined') {
			Post.findById(req.params.id)
			.populate('tags')
			.exec(function(err, post){
				if(err)
					res.send(err);
				
				res.json(post);
			});
		}
		
	}) 
	
	.post(cpUpload, function(req, res) {

	
		Post.findById(req.params.id, function(err, post){
			if(err)
				res.send(err);
			
			// если пользователь удалил старое фото 
			if(req.body.photo == 'null' && post.photo != null) {	
				
					var path = post.photo.path;					
					post.photo = null;					 				
					try {
						fs.unlink(path);
					} catch (err) {								
			       				
					}					
				
			}
			// если пользователь удалил старый трек 
			if(req.body.songs == 'null' && post.songs != null) {
				var songs = post.songs;
				post.songs = null;
				for (var i = 0; i < songs.length; i++) {
					try {
						fs.unlink(songs[i].path);
					} 
					catch (err) {
								       	
					}		
				};			
				
								
			}

			if(req.files) {
				if (req.files['photo_file']) {
					var photo = {};
					photo['url'] = '/img/'+ req.files['photo_file'][0]['filename'];
					photo['path'] = req.files['photo_file'][0]['path'];
					post.photo = photo;
					im.resize({
					  srcPath: photo['path'],
					  dstPath: photo['path'],
					  height:   400,
					  filter: 'Lagrange',
					}, function(err, stdout, stderr){
					  if (err)  console.dir(arguments);
					  im.crop({
						  srcPath: photo['path'],
						  dstPath: photo['path'],
						  width: 400,
						  height: 400,					 
						  
						});
					});
								
				}			
				if (req.files['songs_files']){
					post.songs = [];
					for (var i = 0; i < req.files['songs_files'].length; i++) {
						var song = {
							url : '/sound/'+ req.files['songs_files'][i]['filename'],
							path : req.files['songs_files'][i]['path'],
							name: req.files['songs_files'][i]['originalname']
						};

						post.songs.push(song);					
					};				
				};

			};



			var old_status = post.status;
			post.title = req.body.title;
			post.author = req.body.author;
			post.annotation = req.body.annotation;
			post.content = req.body.content;
			post.last_mod = new Date();
			
			if (req.body.tags != "") {
				post.tags = JSON.parse(req.body.tags);	
			}		

			post.urlname = req.body.urlname;
			post.type = req.body.type;
			
			post.status = req.body.status;

			if (post.status == "publication") {
				if (old_status == "draft") {

					var ester= new Ester();
					ester.post = post._id;
					ester.type = post.type;
					if (req.body.tags != "") {
						ester.tags = JSON.parse(req.body.tags);	
					}
				
					post.ester = ester._id;

					post.save(function(err, post){
						if(err)
							res.send(err);
						ester.save(function(err, ester) {
							if (err)
								return res.send(500, err);						
							return res.send(200, null);
						});
					});
				}
				else {
					Ester.findById(post.ester, function (err, ester) {
						if(err)
							return res.send(500, err);

							post.ester = ester._id;					
							ester.post = post._id;
							ester.type = post.type;
							if (req.body.tags != "") {
								ester.tags = JSON.parse(req.body.tags);	
							}				
							

							post.save(function(err, post){
								if(err)
									res.send(err);
								ester.save(function(err, ester) {
									if (err)
										return res.send(500, err);						
									return res.send(200, null);
								});
							});
					});
				}
				
			}
			else {
				if (old_status == "publication") {
					post.save(function(err, post){
						if(err)
							res.send(err);	
						Ester.findById(post.ester, function (err, ester) {
						if(err)
							return res.send(500, err);
						
						Ester.remove({
							_id: ester._id
						}, function(err) {
							if (err)
								res.send(err);		
							return res.send(200, null);
						});
					});									
											
					});
				}
				else {
					post.save(function(err, post){
						if(err)
							res.send(err);											
						return res.send(200, null);						
					});
				}
			}


			
			
		});
	})
	
	.delete(function(req, res) {

		if (req.params.id) {
				Post.findById(req.params.id, function(err, post){
				if(err)
					res.send(err);				
				if (post.photo != null) {
					try {
						fs.unlink(post.photo.path);
					} catch (err) {								
			       				
					}			
					
				} 
				if (post.songs != null) {
					for (var i = 0; i < post.songs.length; i++) {
						try {
							fs.unlink(post.songs[i].path);
						} 
						catch (err) {
									       	
						}		
				};	
				} 
				Post.remove({
					_id: post._id
				}, function(err) {
					if (err)
						res.send(err);
					if (post.status == "publication") {
						Ester.findById(post.ester, function (err, ester) {
							if(err)
								return res.send(500, err);
							
							Ester.remove({
								_id: ester._id
							}, function(err) {
								if (err)
									res.send(err);		
								res.json("deleted :(");
							});
						});
					}
					else {
						res.json("deleted :(");
					}								
										
				});
					
				
		
			});
			}
			else
				res.send(500, err);

		
	});



router.route('/timetable')	
	
	.post(function(req, res){
		
		var timetable = new Timetable();
		
		timetable.morning = req.body.morning;
		timetable.afternoon = req.body.afternoon;
		timetable.evening = req.body.evening;
		timetable.night = req.body.night;
		var date = new Date(req.body.date);	
		var dt = date.toDateString();
		var newdt = new Date(dt);
		timetable.date = newdt;				
		timetable.date_string = date.getDate().toString() + date.getMonth().toString() + date.getFullYear().toString();
		
		timetable.save(function(err, timetable) {
			if (err){
				return res.send(500, err);
			}
			return res.send(200, null);
		});
	});
	

router.route('/timetable:offset?')	
	.get(function(req, res){
		
		Timetable
			.find()
			.sort({created_at: 'desc'})
			.limit(20)
			.skip( 20 * req.query.offset)
			.select('comment date')
			.sort({date: 'desc'})
			.exec(function (err, timetable) {
				if(err){
					return res.send(500, err);
				}			
				return res.send(200, timetable);
			});
	});

router.route('/timetable/:id')

	.get(function(req, res){
		Timetable.findById(req.params.id, function(err, post){
			if(err)
				res.send(err);
			res.json(post);
		});
	}) 
	
	.put(function(req, res){
		Timetable.findById(req.params.id, function(err, timetable){
			if(err)
				res.send(err);

			timetable.morning = req.body.morning;
			timetable.afternoon = req.body.afternoon;
			timetable.evening = req.body.evening;
			timetable.night = req.body.night;
			var date = new Date(req.body.date);	
			var dt = date.toDateString();
			var newdt = new Date(dt);
			timetable.date = newdt;				
			timetable.date_string = date.getDate().toString() + date.getMonth().toString() + date.getFullYear().toString();		
			
			timetable.save(function(err, timetable){
				if(err)
					res.send(err);

				res.json(200, null);
			});
		});
	})
	
	.delete(function(req, res) {
			if (req.params.id) {
				Timetable.remove({
							_id: req.params.id
						}, function(err) {
							if (err)
								res.send(err);
							res.json("deleted :(");
						});
			}
			else
				res.send(500, err);
		
	});


router.route('/instagram:offset?')
	
	.get(function(req, res){	

	
			Instagram
			.find()
			.sort({created_at: 'desc'})
			.limit(15)
			.skip( 15 * req.query.offset)
			.select('title url created_at content author')			
			.exec(function (err, instagram) {
				if(err){
					return res.send(500, err);
				}			
				return res.send(200, instagram);
			});		
		

	});

router.route('/instagram')	
		
	.post(function(req, res) {

		var instagram = new Instagram();
		var ester= new Ester();

		instagram.title = req.body.title;
		instagram.url = req.body.url;
		instagram.content = req.body.content;
		instagram.author = req.body.author;
		instagram.ester = ester._id,
		ester.instagram = instagram._id;
		ester.type = "instagram";
		if (req.body.tags != "") {
			instagram.tags = req.body.tags;	
			ester.tags = req.body.tags;
		}
		
		instagram
			.save(function(err, instagram) {
			
				if (err)
					return res.send(500, err);	

					ester
						.save(function(err, ester) {
							if (err){
								return res.send(500, err);
							}
							return res.send(200, null);
						});		

			});

	})	
	

router.route('/instagram/:id')	
	
	.delete(function(req, res) {

		if (req.params.id) {
			Instagram
						.findById(req.params.id, function (err, instagram) {
							if(err)
								return res.send(500, err);				
										
							Instagram
								.remove({_id: instagram._id}, function(err) {
										if (err)
											res.send(err);
									  	Ester
									  		.findById(instagram.ester, function (err, ester) {
												if(err){
													return res.send(500, err);
												}			
												Ester
													.remove({_id: ester._id}, function(err) {
														if (err)
															res.send(err);		
												   		res.json("deleted :(");
													});
											});		
								});
					});
		}
		else
			res.send(500, err);


		
		
	});



router.route('/soundcloud:offset?')

	.get(function(req, res){
		
		SoundCloud
			.find()
			.sort({created_at: 'desc'})
			.limit(15)
			.skip( 15 * req.query.offset)
			.exec(function (err, soundCloud) {

				if(err){
					return res.send(500, err);
				}			
				return res.send(200, soundCloud);

			});
	});



router.route('/soundcloud')	
		
	.post(function(req, res) {

		var soundCloud = new SoundCloud();
		var ester= new Ester();

		soundCloud.title = req.body.title;
		soundCloud.url = req.body.url;
		soundCloud.ester = ester._id,
		ester.sound_cloud = soundCloud._id;
		ester.type = "SoundCloud";
		
		if (req.body.tags != "") {
			soundCloud.tags = req.body.tags;	
			ester.tags = req.body.tags;
		}
		console.log(soundCloud);
		console.log(ester);
		soundCloud
			.save(function(err, soundCloud) {
			
				if (err)
					return res.send(500, err);	

					ester
						.save(function(err, ester) {
							if (err){
								return res.send(500, err);
							}
							return res.send(200, null);
						});		

			});

	});	
	
	

router.route('/soundcloud/:id')	
	
	.delete(function(req, res) {
		if (req.params.id) {
			SoundCloud
			.findById(req.params.id, function (err, soundCloud) {
				if(err)
					return res.send(500, err);				
				console.log(soundCloud._id);
				SoundCloud
					.remove({_id: soundCloud._id}, function(err) {
							if (err)
								res.send(err);
						  	Ester
						  		.findById(soundCloud.ester, function (err, ester) {
									if(err){
										return res.send(500, err);
									}			
									Ester
										.remove({_id: ester._id}, function(err) {
											if (err)
												res.send(err);		
									   		res.json("deleted :(");
										});
								});		
					});
		});

		}
		else
			res.send(500, err);
		
		
	});

router.route('/video:offset?')
	.get(function(req, res){
			
			Video
				.find()
				.sort({created_at: 'desc'})
				.limit(15)
				.skip( 15 * req.query.offset)
				.exec(function (err, video) {
					if(err){
						return res.send(500, err);
					}			
					return res.send(200, video);
				});
		});

router.route('/video')	
		
	.post(function(req, res) {

		var video = new Video();
		var ester= new Ester();

		video.title = req.body.title;
		video.url = req.body.url;	
		video.ester = ester._id;

		ester.video = video._id;
		ester.type = "video";
if (req.body.tags != "") {
			video.tags = req.body.tags;	
			ester.tags = req.body.tags;
		}
		video
			.save(function(err, video) {
				if (err){
					return res.send(500, err);
				}
				ester
					.save(function(err, ester) {
						if (err){
							return res.send(500, err);
						}
						return res.send(200, null);
					});	
			});
	}); 
	
	

router.route('/video/:id')	

	.delete(function(req, res) {
		if (req.params.id) {
			Video
			.findById(req.params.id, function (err, video) {
				if(err)
					return res.send(500, err);						
				Video
					.remove({_id: video._id}, function(err) {
							if (err)
								res.send(err);
						  	Ester
						  		.findById(video.ester, function (err, ester) {
									if(err){
										return res.send(500, err);
									}			
									Ester
										.remove({_id: ester._id}, function(err) {
											if (err)
												res.send(err);		
									   		return res.send(200, null);
										});
								});		
					});
		});
		}
		else {
			res.send(500, err);
		}

		
	});

router.route('/tag:offset?')

	.get(function(req, res){		
		Tag
			.find()
			.sort({created_at: 'desc'})
			.limit(30)
			.skip( 30 * req.query.offset)
			.exec(function (err, tag) {
				if(err){
					return res.send(500, err);
				}			
				return res.send(200, tag);
			});
	});



router.route('/tag')	
		
	.post(function(req, res){	
		console.log(req.body);
		var tag = new Tag();
		tag.name = req.body.name;
		tag.urlname = req.body.urlname;
		// tag.weight = req.body.weight.value;		
		console.log(tag);
		tag.save(function(err, tag) {
			if (err){
				return res.send(500, err);
			}
			return res.send(200, null);
		});
	})		
	

router.route('/tag/:id')	
	
	
	.delete(function(req, res) {
	if (req.params.id) {
		Tag.remove({
			_id: req.params.id
		}, function(err) {
			if (err)
				res.send(err);
			res.json("deleted :(");
		});
	} 
	else
		return res.send(500, err);	

		
	});


	router.route('/people')	
		
	.post(peopleUpload, function(req, res) {
		
		var people = new People();	
		people.title = req.body.title;
		people.author = req.body.author;
		people.content = req.body.content;		
		people.urlname = req.body.urlname;			
		people.photo = null;
		people.status = req.body.status;
		people.last_mod = new Date();

		
		if (req.file) {
				var photo = {};
				photo['url'] = '/img/'+ req.file['filename'];
				photo['path'] = req.file['path'];
				people.photo = photo;
				im.resize({
				  srcPath: photo['path'],
				  dstPath: photo['path'],
				  height:   400,
				  filter: 'Lagrange',
				}, function(err, stdout, stderr){
				  if (err)  console.dir(arguments);
				  im.crop({
					  srcPath: photo['path'],
					  dstPath: photo['path'],
					  width: 400,
					  height: 400,					 
					  
					});
				});
							
		};		
				

		people.save(function(err, post) {
			if (err)
				return res.send(500, err);
			res.send(200, null);				
		});

	})	


router.route('/people:title?:offset?')

	.get(function(req, res) {
		if (req.query.title == "") {
			People
			.find()
			.sort({created_at: 'desc'})
			.limit(20)
			.skip( 20 * req.query.offset)
			.select('title status created_at last_mod author')			
			.exec(function (err, people) {
				if(err)
					return res.send(500, err);
				
				res.send(200, people);
			});

		}
		else {

			var a = new RegExp(req.query.title,'gi');
			People
				.find({title: a})
				.sort({created_at: 'desc'})
				.limit(20)
				.skip( 20 * req.query.offset)
				.select('title status created_at last_mod')				
				.exec(function (err, people) {
					if(err)
						return res.send(500, err);
						
					res.send(200, people);
				});
		}
			
		

	});


router.route('/people/:id')

	.get(function(req, res){
		if (req.params.id != 'undefined') {
			People.findById(req.params.id)			
			.exec(function(err, people){
				if(err)
					res.send(err);				
				res.json(people);
			});
		}
		
	}) 
	
	.post(peopleUpload, function(req, res) {

	
		People.findById(req.params.id, function(err, people){
			if(err)
				res.send(err);
			
			// если пользователь удалил старое фото 
			if(req.body.photo == 'null' && people.photo != null) {	
				
					var path = people.photo.path;					
					people.photo = null;					 				
					try {
						fs.unlink(path);
					} catch (err) {								
			       				
					}					
				
			}
			
			
			if (req.file) {
					var photo = {};
					photo['url'] = '/img/'+ req.file['filename'];
					photo['path'] = req.file['path'];
					people.photo = photo;
					im.resize({
					  srcPath: photo['path'],
					  dstPath: photo['path'],
					  height:   400,
					  filter: 'Lagrange',
					}, function(err, stdout, stderr){
					  if (err)  console.dir(arguments);
					  im.crop({
						  srcPath: photo['path'],
						  dstPath: photo['path'],
						  width: 400,
						  height: 400,					 
						  
						});
					});
								
			}			
				
			

			people.title = req.body.title;
			people.author = req.body.author;
			people.content = req.body.content;		
			people.urlname = req.body.urlname;	
			people.status = req.body.status;
			people.last_mod = new Date();			
			
			

			people.save(function(err, people){
				if(err)
					res.send(err);											
					return res.send(200, null);						
			});
		
			
		});
	})
	
	.delete(function(req, res) {

		if (req.params.id) {
			People.findById(req.params.id, function(err, people){
				if(err)
					res.send(err);				
				if (people.photo != null) {
					try {
						fs.unlink(people.photo.path);
					} catch (err) {								
			       				
					}			
					
				} 
				
				People.remove({
					_id: people._id
				}, function(err) {
					if (err)
						res.send(err);					
					res.json("deleted :(");					
				});
					
				
		
			});
		}
		else
			res.send(500, err);

		
	});

router.route('/info')	
	
	.get(function(req, res){
		Info.findOne(function(err, info){
			if (err)
				res.send(500, err);
			res.send(200, info)
		})

	})
		
	.post(function(req, res){	
		console.log(req.body);
			Info.findOne(function(err, info){
			if (err)
				res.send(500, err);

			if (info) {
				info.content = req.body.content;
				info.save(function(err, info){
					if(err)
						res.send(err);
					res.json(200, null);
				});
			}
			else {
				var info = new Info();
				info.content = req.body.content;
				info.save(function(err, info) {
					if (err){
						return res.send(500, err);
					}
					return res.send(200, null);
				});
			}

		});

	})



module.exports = router;
